core/src/stores/postgres.rs

```diff
@@ -1947,19 +1947,42 @@ impl Store for PostgresStore {
             _ => unreachable!(),
         };
 
-        let tx = c.transaction().await?;
+        // Data source documents can be numerous so we want to avoid any transaction that could
+        // potentially hurt the performance of the database. Also we delete documents in small
+        // batches to avoid long running operations.
+        let deletion_batch_size: u64 = 512;
+        let mut total_deleted_rows: u64 = 0;
 
-        let stmt = tx
-            .prepare("DELETE FROM data_sources_documents WHERE data_source = $1")
+        let stmt = c
+            .prepare(
+                "DELETE FROM data_sources_documents WHERE id IN (
+                   SELECT id FROM data_sources_documents WHERE data_source = $1 LIMIT $2
+                 )",
+            )
             .await?;
-        let _ = tx.query(&stmt, &[&data_source_row_id]).await?;
 
-        let stmt = tx.prepare("DELETE FROM data_sources WHERE id = $1").await?;
-        let _ = tx.query(&stmt, &[&data_source_row_id]).await?;
+        loop {
+            // Delete in batches
+            let deleted_rows = c
+                .execute(&stmt, &[&data_source_row_id, &(deletion_batch_size as i64)])
+                .await?;
+            total_deleted_rows += deleted_rows;
+
+            // Break if no more rows to delete
+            if deleted_rows < deletion_batch_size {
+                break;
+            }
+        }
 
-        tx.commit().await?;
+        let stmt = c
+            .prepare("DELETE FROM data_sources_documents WHERE data_source = $1")
+            .await?;
+        let _ = c.query(&stmt, &[&data_source_row_id]).await?;
```

Comment:
why do we need another delete here ? ( if for some reason we still have rows here let's add them to total_deleted_rows )